### PR TITLE
BUG: Add missing check for 0-sized array in ravel_multi_index

### PR DESCRIPTION
**Pull request by seberg**
In wrap and clip modes ravel_multi_index crashes or gave
invalid results if input arrays were empty but the shape not.

---

A small silly fix from 7 years ago I found while trying to clean up my branches a bit.

### CODE REVIEW

File: numpy/lib/tests/test_index_tricks.py [hunk 1]
```diff
@@ -175,6 +175,24 @@ def test_0d(self):
...
+                    np.zeros((3, 1), dtype=np.intp), (2, 1, 0), mode=mode)
+
+    def test_empty_array_unravel(self):
+        res = np.unravel_index(np.zeros(0, dtype=np.intp), (2, 1, 0))
+        # res is a tuple of three empty arrays
+        assert(len(res) == 3)
+        assert(all(a.shape == (0,) for a in res))
+
+        with assert_raises(ValueError):
+            np.unravel_index([1], (2, 1, 0))
```
Comment (mattip) [inline numpy/lib/tests/test_index_tricks.py]: all these tests pass before the fix. Can you write a test that fails before/passes after?

File: numpy/core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -942,6 +942,20 @@ ravel_multi_index_loop(int ravel_ndim, npy_intp *ravel_dims,
...
 
+    /*
+     * Check for 0-dimensional axes unless there is nothing to do.
+     * An empty array/shape cannot be indexed at all.
+     */
+    if (count != 0) {
+        for (i = 0; i < ravel_ndim; ++i) {
+            if (ravel_dims[i] == 0) {
+                PyErr_SetString(PyExc_ValueError,
+                        "cannot unravel if shape has zero entry (is empty).");
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/compiled_base.c]:
```suggestion
                        "cannot unravel if shape has zero entries (is empty).");
```

File: numpy/lib/tests/test_index_tricks.py [hunk 3]
```diff
@@ -175,6 +175,24 @@ def test_0d(self):
...
 
+    @pytest.mark.parametrize("mode", ["clip", "wrap", "raise"])
+    def test_empty_array_ravel(self, mode):
+        res = np.ravel_multi_index(
+                    np.zeros((3, 0), dtype=np.intp), (2, 1, 0), mode=mode)
+        assert(res.shape == (0,))
+
+        with assert_raises(ValueError):
+            np.ravel_multi_index(
+                    np.zeros((3, 1), dtype=np.intp), (2, 1, 0), mode=mode)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_index_tricks.py]: Does this change the error message for `mode="raise"`?

File: numpy/lib/tests/test_index_tricks.py (hunk 1 above)
Comment (seberg) [reply numpy/lib/tests/test_index_tricks.py]: No, I just added those for symmetry with the other function.

File: numpy/lib/tests/test_index_tricks.py (hunk 3 above)
Comment (seberg) [reply numpy/lib/tests/test_index_tricks.py]: The original error (non empty) is: "invalid entry in coordinates array", which is fair. Could just use that, neither is great to be honest.

**Comment by mattip** [comment]
this fixes the segfault from `res = np.ravel_multi_index(np.zeros((3, 1), dtype=np.intp), (2, 1, 0), mode="wrap")`

**Review by hameerabbasi — APPROVED** [review]
LGTM

**Pull request merged**